File: frontend/src/types/mediaAttendance.ts
export interface MediaAttendanceItem {
  student_id?: string | null;
  student_name: string;
  student_code?: string | null;
  roll_number?: string | null;
  confidence: number;
  confidence_pct: number;
  attendance_status: string;
  decision: 'KNOWN' | 'UNKNOWN' | 'UNCERTAIN';
  status?: string;
  first_seen?: string | null;
  last_seen?: string | null;
  observation_count: number;
  remarks?: string | null;
}

export interface UnresolvedFaceItem {
  face_id: string;
  decision: 'UNKNOWN' | 'UNCERTAIN';
  status?: string;
  confidence: number;
  confidence_pct: number;
  timestamp_sec?: number | null;
  frame_number?: number | null;
  bbox: number[];
  quality_score: number;
  rejection_reason?: string | null;
}

export interface DetectedMediaFaceItem {
  face_id: string;
  bounding_box: { x?: number; y?: number; width?: number; height?: number };
  bbox: number[];
  detection_confidence: number;
  quality_score: number;
  identity: string;
  student_id?: string | null;
  student_code?: string | null;
  roll_number?: string | null;
  recognition_confidence: number;
  confidence_pct: number;
  status: 'VERIFIED' | 'VERIFYING' | 'UNKNOWN' | 'LOW_QUALITY';
  rejection_reason?: string | null;
}

export interface MediaAnalysisResponse {
  success?: boolean;
  job_id?: string | null;
  session_id: string;
  session_subject: string;
  session_class: string;
  media_type: 'IMAGE' | 'VIDEO';
  filename: string;
  duration_sec?: number | null;
  resolution?: string | null;
  faces_detected: number;
  recognized_count: number;
  unknown_count: number;
  uncertain_count: number;
  low_quality_count?: number;
  attendance_marked_count: number;
  duplicates_prevented?: number;
  recognized_students: MediaAttendanceItem[];
  unresolved_faces: UnresolvedFaceItem[];
  faces?: DetectedMediaFaceItem[];
  results?: any[];
  processing_time_ms: number;
  status: string;
}

export interface MediaJobResponse {
  id: string;
  session_id: string;
  session_subject?: string | null;
  session_class?: string | null;
  media_type: 'IMAGE' | 'VIDEO';
  filename: string;
  file_size_bytes?: number | null;
  duration_sec?: number | null;
  resolution?: string | null;
  status: 'QUEUED' | 'PROCESSING' | 'COMPLETED' | 'FAILED' | 'CANCELLED';
  progress_pct: number;
  frames_total: number;
  frames_processed: number;
  faces_detected_total: number;
  recognized_count: number;
  unknown_count: number;
  uncertain_count: number;
  attendance_marked_count: number;
  summary_json?: {
    recognized?: MediaAttendanceItem[];
    unresolved?: UnresolvedFaceItem[];
    faces?: DetectedMediaFaceItem[];
    results?: any[];
    duplicates_prevented?: number;
    low_quality_count?: number;
  };
  error_message?: string | null;
  created_by?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
}

export interface SessionBiometricValidationResponse {
  session_id: string;
  subject: string;
  class_name: string;
  status: string;
  total_enrolled_students: number;
  students_with_face_data: number;
  students_missing_face_data: number;
  can_process: boolean;
  warning_message?: string | null;
}

File: backend/app/schemas/media_attendance.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field


class MediaAttendanceItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    student_id: Optional[str] = None
    student_name: str
    student_code: Optional[str] = None
    roll_number: Optional[str] = None
    confidence: float = 0.0
    confidence_pct: float = 0.0
    attendance_status: str = "PRESENT"
    decision: str = "KNOWN"  # KNOWN, UNKNOWN, UNCERTAIN
    status: str = "VERIFIED"  # VERIFIED, VERIFYING, UNKNOWN, LOW_QUALITY
    already_present: bool = False
    alreadyPresent: bool = False
    attendance_marked: bool = True
    attendanceMarked: bool = True
    first_seen: Optional[str] = None
    last_seen: Optional[str] = None
    observation_count: int = 1
    remarks: Optional[str] = None


class UnresolvedFaceItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    face_id: str
    decision: str = "UNKNOWN"  # UNKNOWN, UNCERTAIN
    status: str = "UNKNOWN"  # UNKNOWN, VERIFYING, LOW_QUALITY
    confidence: float = 0.0
    confidence_pct: float = 0.0
    timestamp_sec: Optional[float] = None
    frame_number: Optional[int] = None
    bbox: List[float] = Field(default_factory=list)
    quality_score: float = 1.0
    rejection_reason: Optional[str] = None


class DetectedMediaFaceItem(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    face_id: str
    bounding_box: Dict[str, float] = Field(default_factory=dict, description="{x, y, width, height}")
    bbox: List[float] = Field(default_factory=list, description="[x1, y1, x2, y2]")
    detection_confidence: float = 1.0
    quality_score: float = 1.0
    identity: str = "UNKNOWN"
    student_id: Optional[str] = None
    student_code: Optional[str] = None
    roll_number: Optional[str] = None
    recognition_confidence: float = 0.0
    confidence_pct: float = 0.0
    status: str = "UNKNOWN"  # VERIFIED | VERIFYING | UNKNOWN | LOW_QUALITY
    rejection_reason: Optional[str] = None


class MediaAnalysisResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    success: bool = True
    job_id: Optional[str] = None
    session_id: str
    session_subject: Optional[str] = None
    session_class: Optional[str] = None
    media_type: str = "IMAGE"  # IMAGE, VIDEO
    filename: str
    duration_sec: Optional[float] = None
    resolution: Optional[str] = None
    faces_detected: int = 0
    facesDetected: int = 0
    recognized_count: int = 0
    studentsRecognized: int = 0
    unknown_count: int = 0
    unknownFaces: List[Dict[str, Any]] = Field(default_factory=list)
    uncertain_count: int = 0
    low_quality_count: int = 0
    attendance_marked_count: int = 0
    attendanceMarked: int = 0
    duplicates_prevented: int = 0
    recognized_students: List[MediaAttendanceItem] = Field(default_factory=list)
    attendanceCandidates: List[Dict[str, Any]] = Field(default_factory=list)
    unresolved_faces: List[UnresolvedFaceItem] = Field(default_factory=list)
    faces: List[DetectedMediaFaceItem] = Field(default_factory=list)
    results: List[Dict[str, Any]] = Field(default_factory=list)
    annotatedImagePath: Optional[str] = None
    annotated_image_url: Optional[str] = None
    processing_time_ms: float = 0.0
    status: str = "COMPLETED"
    message: Optional[str] = None


class MediaJobResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    id: str
    session_id: str
    session_subject: Optional[str] = None
    session_class: Optional[str] = None
    media_type: str
    filename: str
    file_size_bytes: Optional[int] = None
    duration_sec: Optional[float] = None
    resolution: Optional[str] = None
    status: str
    progress_pct: float = 0.0
    frames_total: int = 0
    frames_processed: int = 0
    faces_detected_total: int = 0
    recognized_count: int = 0
    unknown_count: int = 0
    uncertain_count: int = 0
    attendance_marked_count: int = 0
    summary_json: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    created_by: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime


class SessionBiometricValidationResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    session_id: str
    subject: str
    class_name: str
    status: str
    total_enrolled_students: int
    students_with_face_data: int
    students_missing_face_data: int
    can_process: bool
    warning_message: Optional[str] = None
